### ISSUE TITLE
average number of [4096,4096] regions per slide

### ISSUE THREAD

**Issue by clemsgrs** [mahmoodlab/HIPT#24]
Hi, 

I'm trying to reproduce your results for the TCGA BRCA dataset. I've thus narrowed down the dataset to the **875** breast slides you use to train & evaluate your models (based on the .csv files you provided [here](https://github.com/mahmoodlab/HIPT/tree/master/2-Weakly-Supervised-Subtyping/splits/10foldcv_subtype/tcga_brca)). 

As a first step, I ran CLAM segmentation & patching pipeline on this dataset. I used the preset parameters that the group provided for TCGA BRCA (can be found [here](https://github.com/mahmoodlab/CLAM/blob/master/presets/tcga.csv)). When computing the average number of regions per slide, I get `avg_M ~ 212`. 

The paper states that the average number of [4096,4096] regions per slide (avg_M) is around 38 when computed over the 10,678 FFPE slides from 33 cancer types in TCGA. I thought "maybe TCGA BRCA slides are much bigger than the other cancer types', hence why such a big difference when computing avg_M". 

To assess whether or not that assumption was true, I downloaded the pre-extracted "region-level" feature embeddings you kindly provide under `3-Self-Supervised-Eval/embeddings_slide_lib/embeddings_slide_lib/vit256mean_tcga_slide_embeddings`. From these I can easily tell how many [4096,4096] regions you found for each of the **875** TCGA BRCA slides. After computing the average over these slides, I get `avg_M ~ 30`.

I was thus wondering if you had an idea of where such a big difference (212 vs. 30) may come from. Do you remember using tcga preset parameters when running CLAM or did you use custom parameters?

Thank you!

**Comment by clemsgrs** [comment]
Oh good call, I wrongly assumed all slides from TCGA came with the same magnification at level 0. 
After checking, the vast majority of them seem to be 40x at level 0, only a few being 20x. Given that I was patching at level 0, it makes sense I had approx. 4 times more patches. I've switched to patching at 20x.

With CLAM preset parameters for TCGA, I now get `avg_M ~ 60`, which is closer your average, though still double the amount. Difference could come from the post-processing you mentioned + different set of parameters.

**Comment by bryanwong17** [comment]
Hi @clemsgrs, May I know to patch at 20x instead of 40x? Is there a value I should change in patch_level?

**Comment by clemsgrs** [comment]
Hi, I've had to write some code in order to patch at 20x instead of 40x.
Also, reasoning with magnification level is not as precise as reasoning with pixel spacing (or resolution).
As said above, most TCGA BRCA slides are 40x at level 0: I inferred this from pixel spacing at level 0 (which is around 0.25 µm/pixel, and most scanners have such a spacing at 40x).

To switch to 20x, I've looked at which level the pixel spacing was closest to 0.50 µm/pixel (twice bigger pixel spacing is equivalent to twice smaller magnification).

My slight adaptation of CLAM pre-processing code should be available [here](https://github.com/clemsgrs/hs2p).
You can take a look at the `get_best_level_for_spacing` function [here](https://github.com/clemsgrs/hs2p/blob/996b530deab91018a91bf0d24fdcebaa10addc15/source/wsi.py#L324).
Just like I did, you only need to call it once at the beginning of the `process_contour` method (see [here](https://github.com/clemsgrs/hs2p/blob/996b530deab91018a91bf0d24fdcebaa10addc15/source/wsi.py#L446))

**Comment by clemsgrs** [comment]
Look at the `get_best_level_for_spacing ` function to find out how I get the pixel spacing at level 0:

```
# OpenSlide gives the resolution in centimeters so we convert this to microns
x_res = float(self.wsi.properties["tiff.XResolution"]) / 10000
y_res = float(self.wsi.properties["tiff.YResolution"]) / 10000
x_spacing, y_spacing = 1 / x_res, 1 / y_res
```

**Comment by bryanwong17** [comment]
@clemsgrs, Got it. I tried to run hs2p, but I got the following error:

Traceback (most recent call last):
  File "E:\hs2p\main.py", line 8, in <module>
    from utils import initialize_wandb, seg_and_patch
  File "E:\hs2p\utils.py", line 12, in <module>
    from source.wsi import WholeSlideImage
  File "E:\hs2p\source\wsi.py", line 6, in <module>
    import openslide
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "<frozen zipimport>", line 259, in load_module
  File "C:\Users\user\anaconda3\lib\site-packages\openslide_python-1.1.2-py3.9-win-amd64.egg\openslide\__init__.py", line 29, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "<frozen zipimport>", line 259, in load_module
  File "C:\Users\user\anaconda3\lib\site-packages\openslide_python-1.1.2-py3.9-win-amd64.egg\openslide\lowlevel.py", line 56, in <module>
  File "C:\Users\user\anaconda3\lib\ctypes\__init__.py", line 460, in LoadLibrary
    return self._dlltype(name)
  File "C:\Users\user\anaconda3\lib\ctypes\__init__.py", line 364, in __init__
    if '/' in name or '\\' in name:
TypeError: argument of type 'NoneType' is not iterable

**Comment by bryanwong17** [comment]
@clemsgrs, Got it. Thank you

from source.wsi_utils import StitchCoords, compute_time

ModuleNotFoundError: No module named 'source.wsi_utils'

I don't see wsi_utils inside of source folder

**Comment by bryanwong17** [comment]
@clemsgrs, Got it. Could you guide me on what should I do now with the below error?

File "C:\Users\user\anaconda3\envs\hipt\lib\site-packages\wandb\sdk\wandb_login.py", line 292, in _login
wlogin.configure_api_key(key)
File "C:\Users\user\anaconda3\envs\hipt\lib\site-packages\wandb\sdk\wandb_login.py", line 176, in configure_api_key
apikey.write_key(self._settings, key)
File "C:\Users\user\anaconda3\envs\hipt\lib\site-packages\wandb\sdk\lib\apikey.py", line 224, in write_key
raise ValueError("API key must be 40 characters long, yours was %s" % len(key))
ValueError: API key must be 40 characters long, yours was 4
Error executing job with overrides: []
Traceback (most recent call last):
File "E:\hs2p\main.py", line 20, in main
wandb_run = initialize_wandb(cfg, key=key)
File "E:\hs2p\utils.py", line 64, in initialize_wandb
group=cfg.wandb.group,
omegaconf.errors.ConfigAttributeError: Key 'group' is not in struct
full_key: wandb.group
object_type=dict

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

**Comment by bryanwong17** [comment]
@clemsgrs, Got it. How to turn off wandb easily? seems like it is connected to all of your code

**Comment by bryanwong17** [comment]
Hi @clemsgrs, I managed to extract 4k patches with 20x (I think). However, they all look the same as only one axis is moving (in my case, only x). Are they normal?

![Screen Shot 2022-12-22 at 12 25 25 AM](https://user-images.githubusercontent.com/67773304/208941110-630524b9-1b23-4375-a579-18b420beea33.png)

**Comment by clemsgrs** [comment]
Hi, have you made sure that the `step_size` parameter in the config file is equal to the `patch_size` parameter?

**Comment by bryanwong17** [comment]
Hi @clemsgrs, thanks for the reminder. I managed to get different patches after changing the step_size parameter.

Based on what you say on this:  > hi, `x` and `y` coordinates should represent the true location in the slide **at level 0**. if pixel spacing is `0.25` at level 0 and you extract [256, 256] patches at spacing `0.5`, then it corresponds to extracting [512, 512] patches at spacing `0.25` ; thus two consecutive patches will have a `512` value difference (either in `x` or `y` coordinates).
> 
> might be better to update coordinates based on spacing, will give it a thought

In this scenario, when I extract [4096,4096] at a spacing 'of 0.5', shouldn't the difference between patches be an '8192' value difference? 

I've modified CLAM code with your slight adaptation, but the difference between patches is 4096. Do I implement it wrongly?

![Screen Shot 2022-12-22 at 8 06 45 AM](https://user-images.githubusercontent.com/67773304/209020068-0a5924d1-ca47-4124-a8ca-ea0bb556c7a0.png)
